### ISSUE TITLE
Error handling for failed bag creation

### ISSUE THREAD

**Issue by bondjimbond** [bondjimbond/islandora_westvault#22]
If while executing the bag-creation loop, one of the bags fails, that kills off the rest of the process:
- Some (but not all?) subsequent objects are not bagged at all
- Subsequent bagged objects do not get their PRESERVATION datastreams updated with <confirmPreserved> tags
- Bagged objects do not get the post_create treatment (moving to new directory)
- Having unmoved files in the /tmp directory (where bags are stored by default before they're moved by post_create) seems to cause problems when objects are re-bagged

This problem sprung up when trying to bag collections; collections apparently cannot be bagged, which triggered all the above problems.

Excluding collections from the process seems to have worked. But what other circumstances might cause bag creation to fail?

Is there a way to read an error/failure and interrupt the rest of the process, so that the offending object can be fixed and the remaining objects don't get messed up?

**Comment by mjordan** [comment]
Islandora Bagit can create Bags for collections. I suggest we figure out why the Bags for collections are failing and try to catch that failure.

But personally I think generating bags at the object level and not the collection level is the way to go, mainly because if you bag a collection, the correction of even a singl-letter typo in metadata of one object means you need to rebag the entire collection. Also, you need to figure out how to handle when/if an object is moved from one collection to another.

**Comment by bondjimbond** [comment]
@mjordan Curious. I tried the "create bag" button on a collection, and got "0 bags created", so assumed that Bag-it doesn't like collections.

I'm quite happy to just ignore collections - but what if the goal is to preserve the collection metadata (rels-ext, mods, dc, etc.)? Is it possible to create a bag for the collection as an object, but not at the same time bagging up its children?

> Also, you need to figure out how to handle when/if an object is moved from one collection to another.

Hmm, a tough one. Is there a share/migrate hook we can use in [islandora.api.php](https://github.com/Islandora/islandora/blob/7.x/islandora.api.php)? I don't see one. Nor one in [islandora_basic_collection.api.php](https://github.com/Islandora/islandora_solution_pack_collection/blob/7.x/islandora_basic_collection.api.php).

**Comment by mjordan** [comment]
If creating collection bags is failing, it's a bug in the Islandora Bagit module. I can't test at the moment but I will do so.

I can't find a hook either. Looking at the code, it appears migrating from one collection to another is a [remove from](https://github.com/Islandora/islandora_solution_pack_collection/blob/7.x/includes/migrate.form.inc#L83) followed by an [add to](https://github.com/Islandora/islandora_solution_pack_collection/blob/7.x/includes/migrate.form.inc#L85) operation, and sharing objects is just a [add to collection](https://github.com/Islandora/islandora_solution_pack_collection/blob/7.x/includes/share.form.inc#L81) opertation. I assume if there were a hook to notify of change in a collection, it would be invoked there. Since collection membership is a relationship, you could probably implement [hook_islandora_datastream_modified](https://github.com/Islandora/islandora/blob/7.x/islandora.api.php#L351) and explicitly check to see if the collection membership relationship has changed.

**Comment by bondjimbond** [comment]
That could work! I'll dig into that after #20 is merged.

**Comment by mjordan** [comment]
On a clean 7.x vagrant, when I click on the Create Bags" tab in a collection, a bag for each child object is generated in my sites/default/files directory:

```
Bag-islandora_15.tgz  Bag-islandora_1.tgz  Bag-islandora_3.tgz  Bag-islandora_5.tgz
Bag-islandora_16.tgz  Bag-islandora_2.tgz  Bag-islandora_4.tgz 
```

No bag for the collection itself is generated (this is how it's designed to work). What are you seeing? Are you expecting a bag for the collection object to be generated as well?

**Comment by bondjimbond** [comment]
Since I was unfamiliar with how Bag-It works, I didn't know what to expect. :)

I'm getting the same as you if I run on a collection that has not already got bags generated -- all children get bagged, collection parent does not.

So I suspect that the problem we were hitting before I had islandora_westvault skip collections was this:
- islandora_westvault sees collection with PRESERVATION datastream, along with all its children
- islandora_westvault runs Bag-It on the collection first, creates bags for all the children
- Post-create does not fire on those child bags (not sure why), so they aren't moved to the new folder
- islandora_westvault attempts to bag the children, sees bags already in /tmp, so does not bother creating new bags for them
- Because they don't get processed, these objects do not get the <confirmPreserved> tag, so they're stuck in a loop to be preserved over and over
- This kills also off the rest of the process, so objects in queue after these children are not processed

> Are you expecting a bag for the collection object to be generated as well?

I'm not sure if people care about doing that or not... since it's not already something Islandora Bagit does, I'm content to leave it alone until people make noise about it. So I'm happy enough to just skip collection objects and only deal with their children.

**Comment by bondjimbond** [comment]
@mjordan I see there's an old Bagit ticket that talks about this... https://jira.duraspace.org/browse/ISLANDORA-1305

> A simple collection object bag, like you said, just containing all of a collection object's datastreams should suffice.

The Collection bag described by Brad is basically the same as what I was thinking as well. But I would additionally like the option to JUST create the Collection bag, without bagging up all the children.

**Comment by bondjimbond** [comment]
Maybe this could be addressed in part with an extra option at https://github.com/Islandora/islandora_bagit/blob/43e4219a6c3086ba3ef284be28cf054f25088a4a/islandora_bagit.module#L147-L156

"Bag Collection object only; children will not be bagged."

**Comment by bondjimbond** [comment]
@mjordan I've partially addressed ISLANDORA-1305 in this branch: https://github.com/bondjimbond/islandora_bagit/tree/ISLANDORA-1305

I created a new "no_children" option for collection bags, which works as hoped for. But it would be good (in terms of addressing the ticket) to also include a collection object bag in the functions [islandora_bagit_create_object_bag_batch() ](https://github.com/bondjimbond/islandora_bagit/blob/ISLANDORA-1305/islandora_bagit.module#L475)and [islandora_bagit_create_collection_bag_batch().](https://github.com/bondjimbond/islandora_bagit/blob/ISLANDORA-1305/islandora_bagit.module#L567)

Perhaps it's just a matter of including the collection PID in the list of objects to include?

Question: Do we have to make it a configurable choice (create a Collection Object bag y/n), or just include it for everyone?

**Comment by bondjimbond** [comment]
> Perhaps it's just a matter of including the collection PID in the list of objects to include?

This might create an infinite loop... more looking required.

**Comment by bondjimbond** [comment]
Things that are needed:
- Confirm bag was created
- Confirm bag actually has the desired pieces (recent folder permissions problem ended up with bags only holding text files, no actual bits from the object were included)
- Only once confirmed should the PRESERVATION datastream be updated

**Comment by bondjimbond** [comment]
@mjordan Can we trust the messages Bagit sends you when a bag is created?

Here's the Bagit output from my drush command:

> vagrant@islandora:/var/www/drupal/sites$ drush -u 1 westvault-bagit
> rename(/tmp/islandora_bagit_tmp/ee2fcde/DC.xml,/tmp/islandora_bagit_tmp/ee2fcde/metadata/dublincore.xml): No such file or directory plugin_object_archivematica_transfers.inc:37                                                 [warning]
> copy(/tmp/islandora_bagit_tmp/ee2fcde/metadata/dublincore.xml): failed to open stream: No such file or directory bagit.php:438                                                                                                   [warning]
> filesize(): stat failed for /tmp/islandora_bagit_tmp/ee2fcde/metadata/dublincore.xml islandora_bagit.module:1144                                                                                                                 [warning]
> rename(/tmp/islandora_bagit_tmp/e209d1d/DC.xml,/tmp/islandora_bagit_tmp/e209d1d/metadata/dublincore.xml): No such file or directory plugin_object_archivematica_transfers.inc:37                                                 [warning]
> copy(/tmp/islandora_bagit_tmp/e209d1d/metadata/dublincore.xml): failed to open stream: No such file or directory bagit.php:438                                                                                                   [warning]
> filesize(): stat failed for /tmp/islandora_bagit_tmp/e209d1d/metadata/dublincore.xml islandora_bagit.module:1144                                                                                                                 [warning]
> Bag created and saved at sites/default/files/Bag-islandora_273.tgz                                                                                                                                                               [status]
> Files added: metadata/dublincore.xml, objects/DC.bin, foo.xml, RELS-EXT.rdf, MODS.xml, DC.bin, PRESERVATION.bin, foxml.xml, PREMIS.xml                                                                                           [status]
> Bag created and saved at sites/default/files/Bag-islandora_18.tgz                                                                                                                                                                [status]
> Files added: metadata/dublincore.xml, objects/DC.bin, foo.xml, RELS-EXT.rdf, MODS.xml, DC.bin, TN.jpg, PRESERVATION.bin, foxml.xml, PREMIS.xml                                                                                   [status]

Those warnings aside (I may want some guidance dealing with those), looks like we get a "Bag created" message and a "Files added" message.

If we can grab that output to analyze, we can review the "files added" message to confirm that all critical files are included. We may need to compare a list of desired critical files with the datastreams on the actual object, though (as not all objects will have an OBJ or PDF datastream - as in the example above).

So, for each object:

- Generate bag
- Analyze bagit output
- Confirm bag is present at new location (Owncloud sync directory)
- IF all of the above is OK, update PRESERVATION datastream; ELSE, don't update datastream and log error

Does that sound good?

**Comment by mjordan** [comment]
> Can we trust the messages Bagit sends you when a bag is created?

When it comes to preservation, and especially pushing stuff into a LOCKSS network (where it is very difficult, by design, to remove content), no, we should not. I'm not being sarcastic, but IMO the best way to 'trust' in this case is to inspect the Bags after they have been created.

Which Islandora BagIt plugins are you recommending/requiring? If it's `plugin_object_ds_basic`, it adds every datastream. If I were you I'd check each Bag for the existence of each datastream. I can help with how to do that.

**Comment by bondjimbond** [comment]
> Which Islandora BagIt plugins are you recommending/requiring? 

We're already pretty strict on required configurations, so maybe `plugin_object_ds_basic` is a good choice... make that the only requirement, and that's the only thing we'll check for.

So that means:

- Confirm a Bag was generated and moved to the correct folder
- Check the object's list of datastreams
- Analyze the bag and compare files vs list
- IF all is well, update PRESERVATION datastream. ELSE log an error and move on.

> I can help with how to do that.

Thanks! I think this will be priority after #36 is tested and merged.

**Comment by mjordan** [comment]
Sounds good.

**Comment by mjordan** [comment]
I think you should be checking to see if `plugin_object_ds_basic` is in fact being used.

**Comment by bondjimbond** [comment]
So, @mjordan... We've got enforcement of `plugin_object_ds_basic` and all the other required configuration thanks to #40.

> If I were you I'd check each Bag for the existence of each datastream. I can help with how to do that.

Once we have that, I think we're sorted. How does one inspect a Bag once it's created?

**Comment by mjordan** [comment]
Inspecting the Bag is not the hard part (just extract the manifest file from it), what's hard is predicting with accuracy what files should be in the manifest independent of the Bag. This task is not specific to WestVault. I'll work on this while I'm away for the next two weeks.

**Comment by mjordan** [comment]
The Bagit library also provides a method for getting a list of files it knows about. That might be a better way to "inspect the Bag" than sucking out the manifest file after the Bag is generated.

**Comment by bondjimbond** [comment]
> what's hard is predicting with accuracy what files should be in the manifest independent of the Bag

Isn't this the point of enforcing the ds_basic plugin? That we have the one file to check and compare against the actual list of files in the bag?

**Comment by bondjimbond** [comment]
Also @mjordan ... do you know of any way we could force the creation of a bad Bag for testing purposes?

**Comment by mjordan** [comment]
One complication I can think of include what to do if more than the basic DS plugin is enabled (e.g., PREMIS plugin or some custom one). Not that difficult, but we'd need to accumulate a list of all the files that all the enabled plugins would produce.

As for creating a bad Bag, we could just delete one of the files from the `/data' directory. Either unzip the serialized Bag and then rezip it, or delete it directly from the zipped Bag using a command-line or GUI tool capable of removing a file from a zip archive.

**Comment by bondjimbond** [comment]
> One complication I can think of include what to do if more than the basic DS plugin is enabled (e.g., PREMIS plugin or some custom one). Not that difficult, but we'd need to accumulate a list of all the files that all the enabled plugins would produce.

Methinks we should start small and get the basic datastreams covered (MVP) and once that feature is built, extend it to the other plugins. Does that make sense?

> As for creating a bad Bag, we could just delete one of the files from the `/data' directory. Either unzip the serialized Bag and then rezip it, or delete it directly from the zipped Bag using a command-line or GUI tool capable of removing a file from a zip archive.

Not really what I meant, unless you were thinking of creating the function outside of islandora_westvault so it could be pointed at a manufactured bad Bag for testing first?

**Comment by mjordan** [comment]
> Methinks we should start small and get the basic datastreams covered (MVP) and once that feature is built, extend it to the other plugins. Does that make sense?

Yes, it does.

Can you be more specific about what you meant about the bad Bag?

**Comment by bondjimbond** [comment]
The idea here is that we're going to check and confirm that the bag does, indeed, contain all the data that we want it to contain.

In order to test that our checks are actually catching bags that are missing expected data, we need to generate bags that are missing expected data so that they can be caught.

**Comment by mjordan** [comment]
Right, so not only does an expected file need to be missing, it also needs to be missing from the manifest.

**Comment by bondjimbond** [comment]
> Right, so not only does an expected file need to be missing, it also needs to be missing from the manifest.

Does it? Does the manifest record what's in the bag, or what _should_ go into the bag?

If it just records what's been put there, it's not very useful for error checking.

**Comment by bondjimbond** [comment]
Or maybe I'm completely misunderstanding; that's quite possible too.

**Comment by mjordan** [comment]
Sorry, I haven't had time to think productively about it and am scrambling to prepare for the iCamp. I'll find some time to think about this a bit more during OR. 

If we validate the Bag first, that will confirm that all files listed in the manifest are present in the /data directory. If the Bag validates, you can trust the manifest. Then, you can compare what is listed in the manifest with what you think should be in the Bag.

**Comment by bondjimbond** [comment]
So is bag validation something within Islandora Bagit?

**Comment by mjordan** [comment]
No, the BagIt library provides a validation method we can call.

**Comment by bondjimbond** [comment]
OK, cool. So then...

>  If the Bag validates, you can trust the manifest. Then, you can compare what is listed in the manifest with what you think should be in the Bag.

What are we checking for, then? That Bagit received the correct instructions? What's the scenario in which a Bag produced via Islandora Westvault will validate but whose manifest will not contain the data we want?

**Comment by bondjimbond** [comment]
@mjordan Have you had a chance to think about this any further? I'm still not clear on how exactly we'd go about validating a Bag.

**Comment by mjordan** [comment]
Have not. Been focused on Islandora 8 leading up to, during, and after the iCamp. I'll try to flesh this out this evening or tomorrow evening. Sorry about the delays.

**Comment by mjordan** [comment]
@bondjimbond I've pushed up to my [fork](https://github.com/mjordan/islandora_westvault-1) a "validate" branch that modifies your drush command to add code to call the BagIt library's `isValid()` method, and then also to get all the filename in the Bag. Note that for the purpose of demonstration, the path to the Bag being validated is hard coded to `/tmp/Bag-Islandora_7.tgz`/

If you run `drush -u 1 westvault-validate-bags` you will get output like this:

```
vagrant@islandora:/var/www/drupal/sites/all/modules/islandora_westvault$ drush -u 1 westvault-validate-bags
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/DC.bin
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/MEDIUM_SIZE.jpg
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/MODS.xml
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/OBJ.jpg
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/RELS-EXT.rdf
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/TECHMD.xml
/tmp/bagit_yaY6bQ/Bag-islandora_7/data/TN.jpg
```

If  `isValid()` is true, then you can trust that all of the files listed are present.

How we check that all of the files *expected* in the Bag are present is tricky. I think you're going to have to create a registry that lists all those files.

Does this help?

**Comment by bondjimbond** [comment]
Hmm... I get no output when I run that Drush command from your branch. (I have an islandora:7 object and the correct filename at the correct path.)

```
vagrant@islandora:/var/www/drupal/sites$ drush -u 1 westvault-validate-bags
vagrant@islandora:/var/www/drupal/sites$ 
```

Am I missing anything?

**Comment by mjordan** [comment]
Don't know. Can you add some debugging prints or something to various places in the command to see how far it gets?

**Comment by bondjimbond** [comment]
Hmm... OK, there was some kind of very weird coincidence that caused this.

It turns out that the `Bag-islandora_7.tgz` in `/tmp` was a directory, not a tgz file. No idea how it got there.

Moved a copy of the true bag to that location. Now if I run the command, it works.

```
vagrant@islandora:/var/www/drupal/sites/all/modules/islandora_westvault$ drush -u 1 westvault-validate-bags
/tmp/bagit_cNuoEW/Bag-islandora_7/data/DC.bin
/tmp/bagit_cNuoEW/Bag-islandora_7/data/MKV.mkv
/tmp/bagit_cNuoEW/Bag-islandora_7/data/MODS.xml
/tmp/bagit_cNuoEW/Bag-islandora_7/data/MP4.mp4
/tmp/bagit_cNuoEW/Bag-islandora_7/data/OBJ.mp4
/tmp/bagit_cNuoEW/Bag-islandora_7/data/PREMIS.xml
/tmp/bagit_cNuoEW/Bag-islandora_7/data/PRESERVATION.bin
/tmp/bagit_cNuoEW/Bag-islandora_7/data/RELS-EXT.rdf
/tmp/bagit_cNuoEW/Bag-islandora_7/data/TECHMD.xml
/tmp/bagit_cNuoEW/Bag-islandora_7/data/TN.jpg
/tmp/bagit_cNuoEW/Bag-islandora_7/data/foo.xml
/tmp/bagit_cNuoEW/Bag-islandora_7/data/foxml.xml
/tmp/bagit_cNuoEW/Bag-islandora_7/data/objects/DC.bin
/tmp/bagit_cNuoEW/Bag-islandora_7/data/objects/OBJ.mp4
```

**Comment by mjordan** [comment]
If you have a list of files that should be in the Bag, can you use the list provided by this sample code to confirm that the expected files are present?

**Comment by bondjimbond** [comment]
From the islandora:7 Solr document: 

```
    "fedora_datastreams_ms": [
        "AUDIT",
        "RELS-EXT",
        "MODS",
        "DC",
        "OBJ",
        "TECHMD",
        "TN",
        "MP4",
        "MKV",
        "PRESERVATION"
    ],
```

So we could capture this Solr field, and confirm that the names of each datastream (except AUDIT) are present in the filenames (minus extensions) in the output of `westvault-validate-bags`, plus confirming the presence of the extra files that aren't datastreams (PREMIS.xml, foxml.xml).

Would that do the job?

**Comment by mjordan** [comment]
It's a good start, but for your sample Bag, you must also have the Archivematica plugin enabled, since your Bag data directory contains an `objects` subdirectory. You might want not only to check the existence of files, but also their expected paths within the `data` directory, like `data/objects/DC.bin`.

**Comment by bondjimbond** [comment]
I pushed a branch based on your branch: https://github.com/bondjimbond/islandora_westvault/tree/validate

I can't seem to get the correct bag path, though; https://github.com/bondjimbond/islandora_westvault/blob/validate/islandora_westvault.drush.inc#L164 does not seem to be the right path. I also tried using the Owncloud path, and it's not there either. So where is the Bag at that stage in the process, so I can target it? (Or is there a simpler way to find it?)

**Comment by bondjimbond** [comment]
Update: turns out it was using the serialized directory instead of the unserialized directory? That doesn't make sense as far as I can tell, but I'll go with it.

Now I'm feeding it the path "sites/default/files/Bag-islandora_175.tgz". Still not working, since this is a relative path. Working on it..

**Comment by bondjimbond** [comment]
There must be a simpler way of getting the Bag's path, though. Any ideas, @mjordan?

**Comment by bondjimbond** [comment]
Update again: Stupid me; forgot to include the ".tgz" extension. Now working...

```
/tmp/owncloud/Bag-islandora_175.tgz
```
Datastreams from Solr:
```
Array
(
    [0] => AUDIT
    [1] => RELS-EXT
    [2] => MODS
    [3] => DC
    [4] => OBJ
    [5] => TECHMD
    [6] => TN
    [7] => JPG
    [8] => JP2
    [9] => PRESERVATION
)
```

Files from the validation function:
```
Array
(
    [0] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/DC.xml
    [1] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/JP2.jp2
    [2] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/JPG.jpg
    [3] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/MODS.xml
    [4] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/OBJ.tif
    [5] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/PREMIS.xml
    [6] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/PRESERVATION.bin
    [7] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/RELS-EXT.rdf
    [8] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/TECHMD.xml
    [9] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/TN.jpg
    [10] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/foo.xml
    [11] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/foxml.xml
    [12] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/metadata/dublincore.xml
    [13] => /tmp/bagit_fhVJHa/Bag-islandora_175/data/objects/OBJ.tif
)
```

**Comment by bondjimbond** [comment]
> It's a good start, but for your sample Bag, you must also have the Archivematica plugin enabled, since your Bag data directory contains an objects subdirectory. You might want not only to check the existence of files, but also their expected paths within the data directory, like data/objects/DC.bin.

I'm not sure how plausible it is to check for every possible plugin that might be selected... Lowest-hanging-fruit is to start by checking if every datastream has a file. Then in a subsequent update can account for the various structures that might be affected by different modules. Does that work?

**Comment by bondjimbond** [comment]
Solved with #41.
